src/Npgsql/NpgsqlRecordDataReader.cs

```diff
@@ -0,0 +1,327 @@
+﻿using Npgsql.Internal;
+using Npgsql.Internal.TypeHandling;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeMapping;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Npgsql
+{
+    internal sealed class NpgsqlRecordDataReader : DbDataReader
+    {
+        NpgsqlDataReader _outermostReader;
+        ulong _uniqueOutermostReaderRowId;
+        int _depth;
+        int _numRows;
+        int _nextRowIndex;
+        int _nextRowBufferPos;
+
+        ColumnInfo[]? _columns;
+
+        bool _calledNextResult;
+        bool _closed;
+
+        struct ColumnInfo
+        {
+            public uint Oid;
+            public int BufferPos;
+        }
+
+        NpgsqlReadBuffer Buffer => _outermostReader.Buffer;
+        ConnectorTypeMapper TypeMapper => _outermostReader.Connector.TypeMapper;
+
+        internal NpgsqlRecordDataReader(NpgsqlDataReader outermostReader, ulong uniqueOutermostReaderRowId, int depth)
+        {
+            _outermostReader = outermostReader;
+            _uniqueOutermostReaderRowId = uniqueOutermostReaderRowId;
+            _depth = depth;
+        }
+
+        internal void InitArray()
+        {
+            var dimensions = Buffer.ReadInt32();
+            var containsNulls = Buffer.ReadInt32() == 1;
+            Buffer.ReadUInt32(); // Element OID. Ignored.
+
+            if (containsNulls)
```

Comment:
Interesting... I wonder if it could make sense to simply skip null rows. But I don't think this is very important/relevant.